### PR TITLE
fix(enhance): reject a factor with more dims than the image instead of hanging

### LINKED ISSUE #4718: enhance: factor broadcasting loop hangs forever when factor has more dims than the image
## Bug

Several `kornia.enhance` functions make `factor` broadcastable with this loop:

```python
while len(factor.shape) != len(image.shape):
    factor = factor[..., None]
```

If `factor` has **more** dimensions than `image`, the loop never ends: each iteration adds a dimension, so the lengths never match. The call hangs forever and never raises.

### Reproduction (current `main`, torch 2.14, CPU)

```python
import torch, kornia
kornia.enhance.adjust_brightness(torch.rand(3, 4, 4), torch.ones(2, 1, 1, 1))  # never returns
```

With a 20 s subprocess timeout, all five of these hang: `adjust_saturation_raw`, `adjust_hue_raw`, `adjust_saturation_with_gray_subtraction`, `adjust_brightness` and `adjust_contrast`. A common trigger is an unbatched `(C, H, W)` image together with a batched factor.

### Occurrences on `main`

```
kornia/enhance/adjust.py:89:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:147:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:219:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:416:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:475:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:550:    while len(factor.shape) != len(image.shape):
kornia/enhance/adjust.py:605:    while len(factor.shape) != len(image.shape):
```

#4717 adds one more copy in the new direct `adjust_saturation` path.

### Suggested fix

Raise a `ValueError` when `factor.dim() > image.dim()`, or when a factor dimension fails to broadcast. Ideally this goes in one shared helper that all the call sites use, with a test per public function.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

### PR DESCRIPTION
**Pull request by mohansree14**
Fixes #4718

### Problem

Seven functions in `kornia/enhance/adjust.py` made `factor` broadcastable with:

```python
while len(factor.shape) != len(image.shape):
    factor = factor[..., None]
```

Each iteration appends a dimension to `factor`, so when `factor` starts with **more** dimensions than `image` the two ranks can never meet and the loop never exits. The call hangs forever rather than raising:

```python
kornia.enhance.adjust_brightness(torch.rand(3, 4, 4), torch.ones(2, 1, 1, 1))  # never returns
```

An unbatched `(C, H, W)` image with a batched factor is the common trigger.

### Fix

All seven call sites now go through one `_make_factor_broadcastable` helper that raises `ValueError` when `factor.dim() > image.dim()`, then pads as before. Fixing it in the shared helper rather than per function means the six siblings that the report did not name are covered too.

The padding still only lines the ranks up. Shapes that pad to an equal rank but do not broadcast keep raising from the op itself, so no currently-working call changes behaviour.

Covered: `adjust_saturation_raw`, `adjust_saturation_with_gray_subtraction`, `adjust_hue_raw`, `adjust_contrast`, `adjust_contrast_with_mean_subtraction`, `adjust_brightness`, `adjust_brightness_accumulative`.

### Tests

`TestFactorBroadcast` in `tests/enhance/test_adjust.py` parametrizes all seven functions over the over-ranked factor and asserts `ValueError`, plus one case pinning that an under-ranked factor still broadcasts. Without the fix that first test hangs (killed at a 20 s timeout); with it, 8 pass.

`tests/enhance` (414 passed, 15 skipped) and the brightness/contrast/saturation/hue/ColorJitter augmentation tests (379 passed) are green locally, as are the `adjust.py` doctests.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `26b6ff2d1`

Thanks, this is a good fix. The shared `_make_factor_broadcastable` helper is the right shape, and it does what #4718 asked for at all seven call sites on this PR's base.

**What I verified**
- Reproduced #4718 on current `main` (`d0b8328e2`). I ran each call in a subprocess with a 12 s timeout, using `(3,4,4)` image + `(2,1,1,1)` factor and `(2,3,4,4)` image + `(2,1,1,1,1)` factor. All nine public functions hang: `adjust_saturation{,_raw,_with_gray_subtraction}`, `adjust_hue{,_raw}`, `adjust_contrast{,_with_mean_subtraction}`, `adjust_brightness{,_accumulative}`. On the PR head all nine raise the new `ValueError`.
- No regression. I compared all 225 factor/image-rank combinations with `factor.dim() <= image.dim()` between `main` and the merge tree: the nine functions × images `(2,3,5,5)`, `(3,5,5)` and `(2,2,3,5,5)` × factors float, 0-d, `(1,)`, `(2,)`, `(2,1,1,1)`, `(1,3,1,1)`, `(3,1,1)`, `(2,2)` and `(3,)`. Output shapes, sums and exception types are identical.
- `tests/enhance/test_adjust.py` passes over float32/64/16/bf16 on the head (526 passed) and on the merge tree (542 passed). The new class collects 32 = 8 × 4 dtypes. The compile subset passes with `KORNIA_TEST_OPTIMIZER=inductor`, the color-augmentation subset passes (77), and so do the `adjust.py` doctests. `dynamo.explain` shows 1 graph and 0 breaks for brightness, contrast and hue with a `(B,)` factor. The rank check runs on static shapes, so the compiled path is unaffected.
- `ValueError` matches the neighbouring shape errors in `adjust.py`, e.g. #4717's "Factor must hold one value per image".
- Merge-tree check against #4722 (RET504 in the same file) is clean in both orders.

**P1: after merging `main`, the public `adjust_saturation` still hangs.** This PR's base predates #4717. #4717 added an 8th copy of the loop in the new direct float32/float64 path of `adjust_saturation` (after the `KORNIA_CHECK`s, before the `factor.shape[-3] != 1` check). The merge is textually clean, so git gives no warning, but the merged file still has one `while len(factor.shape) != len(image.shape):` loop:
```
# merge tree = this PR + origin/main
adjust_saturation(torch.rand(3,4,4), torch.ones(2,1,1,1))       # HANG (>12 s)
adjust_saturation(torch.rand(2,3,4,4), torch.ones(2,1,1,1,1))   # HANG (>12 s)
adjust_saturation_raw(...) / every other function                 # ValueError
```
The new tests stay green on the merge tree because `TestFactorBroadcast` lists only the `_raw`/helper-level functions, not the public `adjust_saturation` and `adjust_hue`. Please merge `main`, route that copy through `_make_factor_broadcastable` and keep the per-channel check after it. Then add `kornia.enhance.adjust_saturation` and `kornia.enhance.adjust_hue` to the parametrization. With float32 in the dtype matrix, that exercises the direct path.

**P2: the equal-rank boundary is not pinned.** A one-token mutation of the helper, `factor.dim() > image.dim()` → `>=`, makes every `(B,1,1,1)` factor and every per-channel `(1,3,1,1)` factor raise. On the merge tree it survives all of `tests/enhance` (443 passed) and the augmentation color/conventions subset (1022 passed). For example, `adjust_brightness(torch.rand(2,3,4,4), torch.full((2,1,1,1), 0.1))` raises `ValueError` under the mutation and works on `main`. `test_factor_with_fewer_dims_still_broadcasts` only asserts the output shape with a `(2,)` factor, so it cannot catch this, and it passes on `main` too. Please add equal-rank cases (`(B,1,1,1)`, and a per-channel `(1,3,1,1)` where the op allows it) with a value oracle, e.g. compare each row against the same op called with that row's float factor.

**P2: changelog fragment.** "Check changelog fragments" is red because there is no fragment. The change is user-visible. On `main` this input never returned a result, so raising breaks nothing that used to work: please add `changelog.d/4720.fixed.md` (not `breaking`). Something like: "`adjust_brightness`, `adjust_contrast`, `adjust_saturation`, `adjust_hue` and their variants now raise `ValueError` for a `factor` with more dimensions than the image, instead of hanging forever (#4718)."

Requesting changes for the merge-tree hang, the unpinned boundary and the fragment. The helper itself needs no changes.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by mohansree14** [comment]
Thanks for the thorough review. All three addressed in 103240ea (on top of a merge of `main`):

**P1** — merged `main` and routed #4717's direct float32/float64 path in `adjust_saturation` through `_make_factor_broadcastable`, keeping the per-channel check after it. No `while len(factor.shape)` loops remain in `kornia/`. `TestFactorBroadcast` now parametrizes the public `adjust_saturation` and `adjust_hue` too, over both image ranks from #4718 (`(3,4,4)` + 4-d factor, `(2,3,4,4)` + 5-d factor). Restoring the old loop makes the public `adjust_saturation` case hang again.

**P2 boundary** — added `test_per_image_factor_with_equal_rank` (all nine ops, `(2,1,1,1)` factor, each row compared against the op called with that row's float) and `test_per_channel_factor_with_equal_rank` (`adjust_brightness` / `adjust_contrast`, `(1,3,1,1)` factor, compared per channel). With the `>` → `>=` mutation, 11 of these fail.

**P2 changelog** — added `changelog.d/4720.fixed.md` with your suggested wording.

`tests/enhance/test_adjust.py`: 630 passed, 25 skipped over float32/64/16/bf16 on CPU. I did not run the inductor compile subset locally; CI should cover it.

**Review by ducha-aiki — APPROVED** [review]
## Re-review at `103240ea1`

Thanks, all three items from the previous round are addressed. Approving.

**P1 (merge-tree hang): fixed.** On the merge tree (`103240ea1` + current `main` 8b03a8085, textually clean), no `while len(factor.shape) != len(image.shape)` loop remains in `kornia/`. The only loop left is inside the helper, behind the rank check. I ran every public entry point in a subprocess with a 30 s timeout, for `(3,4,4)` image + `(2,1,1,1)` factor and for `(2,3,4,4)` image + `(2,1,1,1,1)` factor:
- `adjust_saturation`, `adjust_saturation_raw`, `adjust_saturation_with_gray_subtraction`, `adjust_hue`, `adjust_hue_raw`, `adjust_brightness`, `adjust_brightness_accumulative`, `adjust_contrast`, `adjust_contrast_with_mean_subtraction` and the modules `AdjustSaturation`, `AdjustSaturationWithGraySubtraction`, `AdjustHue`, `AdjustContrast`, `AdjustContrastWithMeanSubtraction`, `AdjustBrightness` and `AdjustBrightnessAccumulative` all raise the new `ValueError`. None hangs.
- `sharpness`, `solarize` and `posterize` already reject this input with their own shape assertions.
- `ColorJiggle` and `ColorJitter` build a `(B,)` factor from their sampled params, so users cannot reach the over-ranked case through them. Their tests pass: tests/augmentation `-k "Color or Brightness or Contrast or Saturation or Hue or Jiggle or Jitter or Gamma"` gives 434 passed, 7 xfailed.

**P2 (boundary): pinned.** I ran each one-token mutant of the fix against `tests/enhance` with `--timeout=20`. All five are killed:

| mutant | result |
|---|---|
| `>` → `>=` | fails (equal-rank tests) |
| `>` → `<` | fails |
| `> image.dim()` → `> image.dim() + 1` | Timeout in `test_factor_with_more_dims_raises[...adjust_saturation]` |
| check dropped (`if False`) | Timeout |
| the loop restored in `adjust_saturation`'s direct float path (the #4717 copy) | Timeout |

The equal-rank tests compare each row or channel against the scalar-factor call. That is a real value oracle, not a shape check, and the tests take the `device`/`dtype` fixtures. `TestFactorBroadcast` collects 160 tests over 4 dtypes.

**P2 (fragment): done.** `changelog.d/4720.fixed.md` is correct. Input that used to hang now raises, so `fixed` is the right type.

**Checks on the merge tree:**
- `tests/enhance/test_adjust.py` over float32/64/16/bf16: 630 passed, 25 skipped.
- `tests/enhance`: 468 passed.
- The inductor dynamo/compile subset of test_adjust passes (9).
- The `adjust.py` doctests pass (26).
- `pre-commit` on the changed files is clean.
- `grep -rnE "#4718|issues/4718" kornia/ tests/ docs/` finds only the fragment.
- CI is 41/41 green on this head.

**Filed, not in this PR's scope:** #4746. `adjust_gamma`/`AdjustGamma` does not hang on an over-ranked `gamma`/`gain`. Its `range(len(input.shape) - len(gamma.shape))` padding goes empty and it broadcasts silently to a larger output, e.g. `(3,4,4)` image + `(2,1,1,1)` gamma → `(2,3,4,4)`. I verified this on `main`. Raising there changes the output for input that currently returns something, so it needs its own `breaking` change.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
